src/wayland/shell/legacy/decoration.rs

```diff
@@ -0,0 +1,128 @@
+    RequestMode {
+        /// KDE decoration object.
+        decoration: OrgKdeKwinServerDecoration,
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+        /// Desired mode.
+        mode: Mode,
+    },
+    /// Decoration object was released.
+    Release {
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+    },
+}
+
+/// Create a new KDE Decoration Manager global
+///
+/// A compositor can set the default mode for new clients at any time using
+/// [`OrgKdeKwinServerDecorationManager::default_mode`]. For individual windows the decorations can
+/// be set using [`OrgKdeKwinServerDecoration::mode`]. Both of these requests will cause the client
+/// to respond with its desired mode using [`KdeDecorationRequest::RequestMode`], which then should
+/// be acknowledged by the compositor using another [`OrgKdeKwinServerDecoration::mode`] event.
+/// **It is up to the compositor to prevent feedback loops**, a client is free to ignore the
+/// suggested decoration mode.
+pub fn init_kde_decoration_manager<L, Impl>(
+    display: &mut Display,
+    implementation: Impl,
+    _logger: L,
+) -> Global<OrgKdeKwinServerDecorationManager>
+where
+    L: Into<Option<slog::Logger>>,
+    Impl: FnMut(KdeDecorationRequest) + 'static,
+{
+    let callback = Rc::new(RefCell::new(implementation));
+    display.create_global(
+        1,
+        Filter::new(
+            move |(decoration_manager, _): (Main<OrgKdeKwinServerDecorationManager>, _), _, _| {
+                (&mut *callback.borrow_mut())(KdeDecorationRequest::Setup {
+                    manager: decoration_manager.deref().clone(),
+                });
+
+                let callback = callback.clone();
+                decoration_manager.quick_assign(move |_, request, _| {
+                    let (id, surface) = match request {
+                        ServerDecorationManagerRequest::Create { id, surface } => (id, surface),
+                        _ => return,
+                    };
+
+                    let decoration = id.deref().clone();
+                    (&mut *callback.borrow_mut())(KdeDecorationRequest::NewToplevelDecoration {
+                        decoration: decoration.clone(),
+                        surface: surface.clone(),
+                    });
+
+                    let callback = callback.clone();
+                    id.quick_assign(move |_, request, _| match request {
+                        ServerDecorationRequest::RequestMode { mode } => {
+                            (&mut *callback.borrow_mut())(KdeDecorationRequest::RequestMode {
+                                decoration: decoration.clone(),
+                                surface: surface.clone(),
+                                mode,
+                            });
+                        }
+                        ServerDecorationRequest::Release => {
```

Comment:
Because the `release` request is a destructor, we can empty this match statement and instead place it in a `id.assign_destructor` or whatever it is called. See xdg-decoration, xdg-foreign and probably a couple other protocols for an example of that.

Reply:
I saw that assign_destructor call but I'm not sure what it does beyond offering a place for implementing custom destructor logic. Is it also required to free the object itself?

Reply:
Pinging @vberger for this question.

Reply:
The main difference is that the closure given to `assign_destructor` is also invoked if the client disconnects without explicitly destroying the objects (such as when forcefully disconnected by a protocol error for example).